tests/mark/check_parametrize.py

```diff
@@ -167,3 +167,190 @@ def function(self, x=1, y=2, z=3):
+
+    def check_defaults_method(self):
+
+        class C(object):
+            @defaults(z=[1, 2])
+            @matrix(x=[1], y=[1, 2])
+            @parametrize(x=3, y=4)
+            def function(self, x=1, y=2, z=-1):
+                return x, y, z
+
+        assert parametrized(C.function)
+        injected_args_list = [m.injected_args for m in C.function.marks]
+        assert len(injected_args_list) == 3
+
+        context_list = MarkedFunctionExpander(None, function=C.function).expand()
+        assert len(context_list) == 6
+
+        expected_output = {
+            (1, 1, 1), (1, 1, 2),
+            (1, 2, 1), (1, 2, 2),
+            (3, 4, 1), (3, 4, 2)
+        }
+        output = set()
+
+        c = C()
+        for ctx in context_list:
+            f = ctx.function
+            injected_args = ctx.injected_args
+            assert f(c) == (injected_args['x'], injected_args['y'], injected_args['z'])
+            output.add(ctx.function(C()))
+
+        assert output == expected_output
+
+    def check_overlap_param_no_override(self):
+
+        @defaults(y=[3, 4], z=[1, 2])
+        @parametrize(w=1, x=2, y=3, z=4)
+        def function(w=10, x=20, y=30, z=40):
+            return w, x, y, z
+
+        assert parametrized(function)
+        injected_args_list = [m.injected_args for m in function.marks]
+        assert len(injected_args_list) == 2
+
+        context_list = MarkedFunctionExpander(None, function=function).expand()
+        assert len(context_list) == 1
+
+        expected_output = {(1, 2, 3, 4)}
+        output = set()
+        for ctx in context_list:
+            output.add(ctx.function())
+
+        assert output == expected_output
+
+    def check_overlap_param_no_override_method(self):
+
+        class C(object):
+            @defaults(y=[3, 4], z=[1, 2])
+            @parametrize(w=1, x=2, y=3, z=4)
+            def function(self, w=10, x=20, y=30, z=40):
+                return w, x, y, z
+
+        assert parametrized(C.function)
+        injected_args_list = [m.injected_args for m in C.function.marks]
+        assert len(injected_args_list) == 2
+
+        context_list = MarkedFunctionExpander(None, function=C.function).expand()
+        assert len(context_list) == 1
+
+        expected_output = {(1, 2, 3, 4)}
+        output = set()
+        for ctx in context_list:
+            output.add(ctx.function(C()))
+
+        assert output == expected_output
+
+    def check_overlap_matrix_no_override(self):
+
+        @defaults(y=[3, 4], z=[1, 2])
+        @matrix(x=[1, 2], y=[5, 6])
+        def function(x=20, y=30, z=40):
+            return x, y, z
+
+        assert parametrized(function)
+        injected_args_list = [m.injected_args for m in function.marks]
+        assert len(injected_args_list) == 2
+
+        context_list = MarkedFunctionExpander(None, function=function).expand()
+        assert len(context_list) == 4
+
+        expected_output = {(1, 5, 40), (1, 6, 40), (2, 5, 40), (2, 6, 40)}
```

Comment:
this seems a bit confusing. the default value and `@defaults` annotation just seem at odds with each other, is this really the result we want? i think if i was just reading the code in a test I would actually expect the `z` parameterizations from `@defaults` to be used, but i'm not sure either one is really right

Reply:
this will change once I change the logic above